fendermint/vm/topdown/src/lib.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+mod cache;
+mod error;
+mod finality;
+mod sync;
+
+use async_stm::StmResult;
+use ipc_agent_sdk::message::ipc::ValidatorSet;
+use ipc_sdk::cross::CrossMsg;
+use serde::{Deserialize, Serialize};
+
+pub use crate::cache::{SequentialAppendError, SequentialKeyCache, ValueIter};
+pub use crate::error::Error;
+pub use crate::finality::InMemoryFinalityProvider;
+pub use crate::sync::{IPCAgentProxy, PollingParentSyncer};
+
+type BlockHeight = u64;
+type Bytes = Vec<u8>;
+type BlockHash = Bytes;
+
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The number of blocks to delay before reporting a height as final on the parent chain.
+    /// To propose a certain number of epochs delayed from the latest height, we see to be
+    /// conservative and avoid other from rejecting the proposal because they don't see the
+    /// height as final yet.
+    pub chain_head_delay: BlockHeight,
+    /// Parent syncing cron period, in seconds
+    pub polling_interval_secs: u64,
+}
+
+/// The finality view for IPC parent at certain height.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct IPCParentFinality {
+    /// The latest chain height
+    pub height: BlockHeight,
+    /// The block hash. For FVM, it is a Cid. For Evm, it is bytes32 as one can now potentially
+    /// deploy a subnet on EVM.
+    pub block_hash: BlockHash,
+}
+
+pub trait ParentViewProvider {
+    /// Get the latest height of the parent recorded
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>, Error>;
```

Comment:
I was deciding on completely using `Result` or just `Option`, but then I reckoned that `Result` is better because there are different scenarios that it could result in `Error` and upstream might want to handle the different scenarios. Just returning `None` might not be able to differentiate the scenarios of `None`. In the end `StmResult<Option<BlockHeight>, Error>` is chosen because if the parent has stopped producing block in someway and parent view has caughted up, so there wont be any new `height`, I don't consider this as an error, but a `None`. That's why it's `StmResult<Option<BlockHeight>, Error>`.

Reply:
But this is the provider, it's there to provide a view to the parent. It was absolutely not clear anywhere that if I ask "what's the latest height on the parent that you know of" it might return `None` even after it has returned `Some` before because it considers the parent stopped producing blocks. Why complicate it like this? How long must the parent be dead? Why do we consider this scenario at all? Why not let the caller decide that `latest_height` has not moved for a long time?

Reply:
Now that I see that the parent view only has finalized blocks, and it also removes them after they have been executed, I feel this whole interface is rather misleading. `latest_height` is really `latest_finalized_height` _but_ it's `None` after we execute, so it's like the `next_finalized_height`, but that's exactly what the `next_proposal` (forgot the name) is after, so maybe it would be easier to just hide all these details and reduce them to the three methods that the interpreter actually needs to call (`next`, `check`) and the `new` which is called to extend it. Even the syncer can use `next` to figure out where to resume fetching from.